lib/rules/declaration-block-no-duplicate-properties/index.js

```diff
@@ -88,97 +88,84 @@ const rule = (primary, secondaryOptions, context) => {
 
 				const indexDuplicate = decls.findIndex((d) => d.prop.toLowerCase() === lowerProp);
 
-				if (indexDuplicate !== -1) {
-					const duplicateDecl = decls[indexDuplicate];
-					const duplicateValue = duplicateDecl ? duplicateDecl.value : '';
-					const duplicateImportant = duplicateDecl ? duplicateDecl.important : false;
-
-					if (ignoreDiffValues || ignorePrefixlessSameValues) {
-						// fails if duplicates are not consecutive
-						if (indexDuplicate !== decls.length - 1) {
-							if (context.fix) {
-								if (!important && duplicateImportant) {
-									decl.remove();
-								} else {
-									removePreviousDuplicate(decls, lowerProp);
-								}
-
-								return;
-							}
+				if (indexDuplicate === -1) {
+					decls.push(decl);
+				}
+
+				const duplicateDecl = decls[indexDuplicate];
+
+				if (!duplicateDecl) {
+					return;
+				}
+
+				const duplicateValue = duplicateDecl.value || '';
+				const duplicateImportant = duplicateDecl.important || false;
+				const duplicateIsMoreImportant = !important && duplicateImportant;
+				const duplicatesAreConsecutive = indexDuplicate === decls.length - 1;
 
-							report({
+				if (ignoreDiffValues || ignorePrefixlessSameValues) {
+					if (!duplicatesAreConsecutive) {
+						if (!context.fix) {
+							return report({
 								message: messages.rejected(prop),
 								node: decl,
 								result,
 								ruleName,
 								word: prop,
 							});
-
-							return;
 						}
 
-						if (ignorePrefixlessSameValues) {
-							// fails if values of consecutive, unprefixed duplicates are equal
-							if (vendor.unprefixed(value) !== vendor.unprefixed(duplicateValue)) {
-								if (context.fix) {
-									if (!important && duplicateImportant) {
-										decl.remove();
-									} else {
-										removePreviousDuplicate(decls, lowerProp);
-									}
+						if (duplicateIsMoreImportant) {
+							return decl.remove();
+						}
 
-									return;
-								}
+						return duplicateDecl.remove();
+					}
 
-								report({
+					if (ignorePrefixlessSameValues) {
+						// fails if values of consecutive, unprefixed duplicates are equal
+						if (vendor.unprefixed(value) !== vendor.unprefixed(duplicateValue)) {
+							if (!context.fix) {
```

Comment:
It seems like you have
```
if (!context.fix) {
	return report({
		message: messages.rejected(prop),
		node: decl,
		result,
		ruleName,
		word: prop,
	});
}

if (duplicateIsMoreImportant) {
	return decl.remove();
}

return duplicateDecl.remove();
```
twice.

What about

```js
// false if values of consecutive, unprefixed duplicates are equal
const foo = vendor.unprefixed(value) !== vendor.unprefixed(duplicateValue;
const bar = ignorePrefixlessSameValues && foo;

if (!duplicatesAreConsecutive || bar) {
  …
}
```
at line 107?

Reply:
Not sure how to name `bar` really. I guess, creating `unprefixedDuplicatesAreEqual` var is enough

Reply:
Also thought of moving repetitive fix logic to a separate function, what do you think?

Reply:
I am just making suggestions, you are the one who instigated that refactor, hence you should know better than me.

Reply:
I am just making suggestions, you are the one who instigated that refactor, hence you should know better than me.

All I can say is that
```
  ruleName,
  word: prop,
```
is currently missing after l141.
If the reports' arguments were identical you could have an argument-less function for both.
If the function requires arguments, I would advise against it.